scripts/tasks/preprocess-api-history.ts

```diff
@@ -0,0 +1,231 @@
+import globby from 'globby';
+import path from 'path';
+import { fromMarkdown } from 'mdast-util-from-markdown';
+import { visit } from 'unist-util-visit';
+import type { Html, Root } from 'mdast';
+import type { Node, Literal } from 'unist';
+import Ajv, { JSONSchemaType, ValidateFunction } from 'ajv';
+import { readFile, writeFile } from 'fs/promises';
+import { fromHtml } from 'hast-util-from-html';
+import logger from '@docusaurus/logger';
+import { parse as parseYaml } from 'yaml';
+
+const apiHistoryRegex = /<!--[\s\S]*?(```[\s\S]*?```)[\s\S]*?-->/;
+
+interface ChangeSchema {
+  'pr-url': string;
+  'breaking-changes-header'?: string;
+  description?: string;
+}
+
+interface ApiHistory {
+  added?: ChangeSchema[];
+  deprecated?: ChangeSchema[];
+  changes?: ChangeSchema[];
+}
+
+interface LiteralString extends Literal {
+  value: string;
+}
+
+// Copied from here: <https://github.com/electron/website/blob/feat/api-history/scripts/tasks/add-frontmatter.ts#L16-L23>
+const getMarkdownFiles = async (startPath: string) => {
+  const filesPaths = await globby(path.posix.join(startPath, 'api', '/*.md'));
+
+  const files: Map<string, string> = new Map();
+  for (const filePath of filesPaths) {
+    const content = await readFile(filePath, 'utf-8');
+    files.set(filePath, content);
+  }
+
+  return files;
+};
+
+// Copied from here but slightly different: <https://github.com/electron/lint-roller/blob/bac245478ba69017c2a82e8ba1b2884a37647494/bin/lint-markdown-api-history.ts#L42-L67>
+function findPossibleApiHistoryBlocks(tree: Root) {
+  const codeBlocks: Html[] = [];
+
+  visit(
+    tree,
+    (node: Node) =>
+      node.type === 'html' &&
+      (node as LiteralString).value.includes('```') &&
+      (node as LiteralString).value.toLowerCase().includes('yaml') &&
+      (node as LiteralString).value.toLowerCase().includes('history'),
+    (node) => {
+      codeBlocks.push(node as Html);
+    }
+  );
+
+  return codeBlocks;
+}
+
+function findValidApiHistoryBlocks(
+  possibleHistoryBlocks: Html[],
+  filePath: string,
+  validateAgainstSchema: ValidateFunction<ApiHistory>
+) {
+  const validHistoryBlocks: Html[] = [];
+
+  // All of the validation logic copied from here: <https://github.com/electron/lint-roller/blob/bac245478ba69017c2a82e8ba1b2884a37647494/bin/lint-markdown-api-history.ts#L117-L176>
+  for (const possibleHistoryBlock of possibleHistoryBlocks) {
```

Comment:
For all of the `continue` clauses in this loop, can we add some line/column numbers based on `possibleHistoryBlock.position`?

I imagine it would be hard to find the exact failure spot for an API file as big as `app.md` for example.

Reply:
Isn't this script meant to just check the block is valid and process it? If you want more details aren't you meant to use the `lint-roller` script?

Reply:
Discussed in Slack meeting, resolved in [`04029d1` (#594)](https://github.com/electron/website/pull/594/commits/04029d12875b8862a2139c1839a107f7df1c318d)

Reply:
As discussed in the sync today, we should do two things to make the error message clearer:

* Clarify that the `Error` will simply skip the block rather than fail parsing.
* Tell users to run `lint-roller` to resolve any warnings.

Reply:
> As discussed in the sync today, we should do two things to make the error message clearer:
> 
> * Clarify that the `Error` will simply skip the block rather than fail parsing.
> * Tell users to run `lint-roller` to resolve any warnings.

Sorry, was this a delayed reply? I think [`04029d1` (#594)](https://github.com/electron/website/pull/594/commits/04029d12875b8862a2139c1839a107f7df1c318d) implements what you mentioned here.